wasmplugin/plugin.go

```diff
@@ -169,6 +168,26 @@ func (ctx *httpContext) OnHttpRequestHeaders(numHeaders int, endOfStream bool) t
 		return ctx.handleInterruption("http_request_headers", interruption)
 	}
 
+	contentLen, err := proxywasm.GetHttpRequestHeader("content-length")
+	if err != nil || contentLen == "0" {
```

Comment:
This is the very important point @mathetake @jptosso @fzipi @M4tteoP

Reply:
I think we should read the RFC here. `Transfer-Encoding` will modify this behavior. https://datatracker.ietf.org/doc/html/rfc7230#section-3.3.2

Reply:
Interestingly, when we send 'Transfer-Encoding' header, not only it reaches upstream directly but also the header `Transfer-Encoding` does not get to upstream and none of the proxy-wasm callbacks are called.

<img width="740" alt="image" src="https://user-images.githubusercontent.com/3075074/210566234-7121c01e-97b3-43e2-8c47-4fde7050aa50.png">

Reply:
Chunked encoding means read chunks until final `\r\n`, so that is minimum for it to not hang. Presumably a payload with only that would also not call our lifecycle method though. There's no way to detect this with headers.

Instead we should backfill, not prefill. We can add `lastPhaseRun int` field and `runPhasesTo` method and use that in each lifecycle method (including destroy) instead of a single phase. That should solve it without worrying about HTTP semantics - it will also allow Envoy direct responses to pass through the filter which currently doesn't happen, while such responses are generally trustworthy it will be good to have it working anyways.

Reply:
Actually I realized such logic could be in coraza itself, with the tracker on `Transaction`. That seems like a good place for it